engine/standalone/engine/src/main/scala/pl/touk/nussknacker/engine/standalone/StandaloneProcessInterpreter.scala

```diff
@@ -141,36 +141,83 @@ object StandaloneProcessInterpreter {
     private def partInvoker(node: compiledgraph.node.Node, parts: List[SubsequentPart]): CompilationResult[WithSinkTypes[InterpreterType]] = {
 
       compilePartInvokers(parts).map(_.map { partsInvokers =>
-        (ctx: Context, ec: ExecutionContext) => {
+        (ctx: List[Context], ec: ExecutionContext) => {
           implicit val iec: ExecutionContext = ec
           //TODO: refactor StandaloneInterpreter to use IO
           implicit val shape: InterpreterShape[Future] = new FutureShape
-          processCompilerData.interpreter.interpret[Future](node, processCompilerData.metaData, ctx).flatMap { maybeResult =>
-            maybeResult.fold[InterpreterOutputType](
-            ir => Future.sequence(ir.map(interpretationInvoke(partsInvokers))).map(foldResults),
-            a => Future.successful(Left(NonEmptyList.of(a))))
+          Future.sequence(ctx.map { one =>
+            processCompilerData.interpreter.interpret[Future](node, processCompilerData.metaData, one)
+          }).flatMap { list =>
+            val foldedResults = foldResults(list.map(_.swap.leftMap(NonEmptyList.one)))
+            val interpreterd = foldedResults.map { resultList =>
+              resultList.groupBy(_.reference).map {
+                case (pr, ir) => interpretationInvoke(partsInvokers)(pr, ir)
+              }.toList
+            }
+            interpreterd match {
+              case Right(other) => Future.sequence(other).map(foldResults)
```

Comment:
why not to move `Future.sequence(other).map(foldResults)` into mapping of `foldedResults`? It will be visible then that you here only transforming either[future] into future[either]. Also maybe we should use for-comprehension for futures and/or extract method for all this folding to make it more human readable?

Reply:
I refactored a bit, should be clearer a bit now